UE4Project/Plugins/AirSim/Source/RenderRequest.cpp

```diff
@@ -3,181 +3,57 @@
-            for (unsigned int i = 0; i < req_size_; ++i) {
-                params_[i]->render_component->CaptureSceneDeferred();
-            }
+    UAirBlueprintLib::RunCommandOnGameThread([this]() {
+        fast_cap_done_ = false;
+        fast_rt_resource_ = fast_param_.render_component->TextureTarget->GameThread_GetRenderTargetResource();
+        fast_param_.render_component->CaptureScene();
+        ENQUEUE_RENDER_COMMAND(SceneDrawCompletion)([this](FRHICommandListImmediate& RHICmdList)
+        {
+            this->RenderThreadScreenshotTask(this->latest_result_);
         });
+    }, false);
 
-        // wait for this task to complete
-        while (!wait_signal_->waitFor(5)) {
-            // log a message and continue wait
-            // lamda function still references a few objects for which there is no refcount.
-            // Walking away will cause memory corruption, which is much more difficult to debug.
-            UE_LOG(LogTemp, Warning, TEXT("Failed: timeout waiting for screenshot"));
-        }
-    }
-
-    for (unsigned int i = 0; i < req_size; ++i) {
-        if (!params[i]->pixels_as_float) {
-            if (results[i]->width != 0 && results[i]->height != 0) {
-                results[i]->image_data_uint8.SetNumUninitialized(results[i]->width * results[i]->height * 3, false);
-                if (params[i]->compress)
-                    UAirBlueprintLib::CompressImageArray(results[i]->width, results[i]->height, results[i]->bmp, results[i]->image_data_uint8);
-                else {
-                    uint8* ptr = results[i]->image_data_uint8.GetData();
-                    for (const auto& item : results[i]->bmp) {
-                        *ptr++ = item.B;
-                        *ptr++ = item.G;
-                        *ptr++ = item.R;
-                    }
-                }
-            }
-        }
-        else {
-            results[i]->image_data_float.SetNumUninitialized(results[i]->width * results[i]->height);
-            float* ptr = results[i]->image_data_float.GetData();
-            for (const auto& item : results[i]->bmp_float) {
-                *ptr++ = item.R.GetFloat();
-            }
-        }
-    }
-}
-
-FReadSurfaceDataFlags RenderRequest::setupRenderResource(const FTextureRenderTargetResource* rt_resource, const RenderParams* params, RenderResult* result, FIntPoint& size)
-{
-    size = rt_resource->GetSizeXY();
-    result->width = size.X;
-    result->height = size.Y;
-    FReadSurfaceDataFlags flags(RCM_UNorm, CubeFace_MAX);
-    flags.SetLinearToGamma(false);
-
-    return flags;
+    //Try to wait just long enough for the render thread to finish the capture.
+    //Start with a long wait, then check for completion more frequently.
+    //TODO: Optimize these numbers.
+    for (int best_guess_cap_time_microseconds = 500; !fast_cap_done_; best_guess_cap_time_microseconds = 200)
+        std::this_thread::sleep_for(std::chrono::microseconds(best_guess_cap_time_microseconds));
```

Comment:
I guess we want to tweak these numbers for our specific hardware?

Reply:
Hmmm good catch.
Unfortunately there is too huge of a delay between running tests, since the simulator is very slow to start on the gcp instances. Still if we have some time later on it is definitly worth a try to tweak those.